swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -148,11 +148,34 @@ private module Cached {
     // flow through `!`
     nodeFrom.asExpr() = nodeTo.asExpr().(ForceValueExpr).getSubExpr()
     or
-    // flow through `?`
+    // flow through `?` and `?.`
     nodeFrom.asExpr() = nodeTo.asExpr().(BindOptionalExpr).getSubExpr()
     or
     nodeFrom.asExpr() = nodeTo.asExpr().(OptionalEvaluationExpr).getSubExpr()
     or
+    // flow through nil-coalescing operator `??`
+    exists(BinaryExpr nco |
+      nco.getFunction().(DeclRefExpr).getDecl().(FreeFunctionDecl).getName() = "??(_:_:)" and
+      (
+        // value argument
+        nodeFrom.asExpr() = nco.getAnOperand()
+        or
+        // unpack closure (the second argument is typically an `AutoClosureExpr` argument)
+        nodeFrom.asExpr() =
+          nco.getAnOperand().(AbstractClosureExpr).getBody().getAnElement().(ReturnStmt).getResult()
+      ) and
+      nodeTo.asExpr() = nco
+    )
```

Comment:
A couple of cleanup things. I _think_ the second operand of `??` will always be an `AutoClosureExpr`. In that case, we can use the `getReturn` predicate on that class. There are two additional cleanups we can do here:
- Calling `getStaticTarget` on a `BinaryExpr` is a bit weird. Maybe we should have a synonym (like `getOperator`?) defined on `BinaryExpr` (and similarly for `UnaryExpr`s).
- `AutoClosureExpr.getReturn()` is necessary in CFG construction, but non-CFG  code probably wants a way to get the _expression_ being returned (instead of getting the `ReturnStmt`). Maybe we should have a `getExpr` predicate on `AutoClosureExpr` that's defined as `this.getReturn().getResult()`?

```suggestion
  exists(BinaryExpr nco |
    nco.getStaticTarget().(FreeFunctionDecl).getName() = "??(_:_:)" and
    nodeTo.asExpr() = nco
  |
    // value argument
    nodeFrom.asExpr() = nco.getAnOperand()
    or
    // unpack closure (the second argument is typically an `AutoClosureExpr` argument)
    nodeFrom.asExpr() = nco.getAnOperand().(AutoClosureExpr).getReturn().getResult()
  )
```

Reply:
> I think the second operand of `??` will always be an `AutoClosureExpr`.

My understanding is that you can provide a manual closure, in which case it won't be an `AutoClosureExpr`.  I could be wrong.  The `@autoclosure` tag on the second argument just means (I think) that Swift is allowed to automatically turn an expression in this position into a closure (`AutoClosureExpr`).  So you can write:
```
x ?? 0
```
rather than:
```
x ?? { return 0 }
```
or something like that.

---

I'll add a `BinaryExpr.getOperator` (or something similar, maybe `.getOperation`?).

---

I agree about `AutoClosureExpr.getExpr` but I don't think I want to use `AutoClosureExpr` in this PR.

Reply:
> My understanding is that you can provide a manual closure, in which case it won't be an `AutoClosureExpr`.

... every doc I read backs me up that `@autoclosure` "lets you omit braces" or similar phrasing, but `swiftc` is not happy if I leave them in.  So it appears you're right, `@autoclosure` actually means you *must* omit the braces, and you presumably always get an `AutoClosureExpr` as a result.

Reply:
All points addressed, I think.  There's no test coverage for the unary expr predicates though.